pkg/event/dispatcher.go

```diff
@@ -94,24 +95,23 @@ type QueueEventDispatcher struct {
 // DispatchEvent queues event with callback and calls flush in a go routine.
 func (ed *QueueEventDispatcher) DispatchEvent(event LogEvent) (bool, error) {
 	ed.eventQueue.Add(event)
-	go func() {
-		ed.flushEvents()
-	}()
+	go ed.flushEvents()
 	return true, nil
 }
 
 // flush the events
 func (ed *QueueEventDispatcher) flushEvents() {
 
-	ed.eventFlushLock.Lock()
-
-	defer func() {
-		ed.eventFlushLock.Unlock()
-	}()
+	// Limit flushing to a single worker
+	if !ed.processing.TryAcquire(1) {
+		return
+	}
+	defer ed.processing.Release(1)
 
 	retryCount := 0
-	ed.queueSize.Set(float64(ed.eventQueue.Size()))
-	for ed.eventQueue.Size() > 0 {
+	queueSize := ed.eventQueue.Size()
+	for ; queueSize > 0; queueSize = ed.eventQueue.Size() {
```

Comment:
I think I understand, but can you provide a quick comment explaining why this has to be in the loop?

Reply:
Since the Size() was only evaluated at the beginning and the end of the loop. We failed to capture any queue size that may have evolved while the loop was being processed. Under heavy load, the queue can grow and this loop will continue to iterate, but the size would not be captured. I discovered this while investigating runaway memory allocation. Our metrics were reporting a constant size in the single digits, but in reality it was growing into the hundreds of thousands.